src/CloudFoundry.WinDEA/StagingInstance.cs

```diff
@@ -265,34 +267,226 @@ public void GetBuildpack(StagingStartMessageRequest message, string gitPath, str
+
+                        var buildpackMutex = new Mutex(false, "admin-buildpack-" + ab.Key);
+                        try
+                        {
+                            buildpackMutex.WaitOne();
+
+                            DownloadAdminBuildpack(ab, adminBuildpackDir);
+
+                            var winBuildpack = IsWindowsCompatibleBuildpack(dir);
+                            if (!winBuildpack)
+                            {
+                                Logger.Info("Staging task {0}. Unable to detect compatible windows executables in buildpack bin. Ignoring buildpack {1}", this.Properties.TaskId, dir);
+                                continue;
+                            }
+
+                            DEAUtilities.DirectoryCopy(dir, Path.Combine(this.Workspace.TempDir, "buildpack"), true);
+                        }
+
+                        finally
+                        {
+                            buildpackMutex.ReleaseMutex();
+                            buildpackMutex.Dispose();
+                        }
+
+                        Buildpack bp = new Buildpack(Path.Combine(this.Workspace.TempDir, "buildpack"), Path.Combine(this.Workspace.StagedDir, "app"), this.Workspace.Cache, this.Workspace.StagingLogPath);
+                        bool success = bp.Detect(this.Container);
+                        if (success)
+                        {
+                            this.Buildpack = bp;
+                            break;
+                        }
+                        else
+                        {
+                            Directory.Delete(Path.Combine(this.Workspace.TempDir, "buildpack"), true);
+
+                        }
                     }
-                    else
+                }
+
+                if (this.Buildpack == null)
+                {
+                    foreach (string dir in systemBuildpacks)
                     {
-                        Directory.Delete(Path.Combine(this.Workspace.TempDir, "buildpack"), true);
+                        DEAUtilities.DirectoryCopy(dir, Path.Combine(this.Workspace.TempDir, "buildpack"), true);
+                        Buildpack bp = new Buildpack(Path.Combine(this.Workspace.TempDir, "buildpack"), Path.Combine(this.Workspace.StagedDir, "app"), this.Workspace.Cache, this.Workspace.StagingLogPath);
+                        bool success = bp.Detect(this.Container);
+                        if (success)
+                        {
+                            this.Buildpack = bp;
+                            break;
+                        }
+                        else
+                        {
+                            Directory.Delete(Path.Combine(this.Workspace.TempDir, "buildpack"), true);
+                        }
                     }
                 }
 
+
                 if (this.Buildpack == null)
                 {
                     throw new Exception("Unable to detect a supported application type");
```

Comment:
InvalidOperationException please. Also, could we add more information about the application in this message?